Normalization + Population scripts and data

HotLemonJuice opened the pull request:
Related issue: #156 

Objectives: 
- Show normalized count (i.e. count / population) using lookup query

Changes made:
- `map/controller.js` for UI and `common/service.js` for lookup
- Included state/county/city population data, crawler, and data-cleaning scripts in `noah/src/main/resources/population`

Implementation detail:
- Use Bootstrap Toggle framework to add the toggle in the front end
- Use lookup inside `byGeoRequest` in `service.js` for population data; depends on the `geolevel`, it will query a different dataset from `dsStatePopulation`/`dsCountyPopulation`/`dsCityPopulation`
- Redraw the map whenever the normalization toggle is toggled; by default, the normalization is turned off
- Crawl city/county data using Scrapy (a Python framework) from https://www.citypopulation.de/php/usa-census-alabama.php and state data from https://www.citypopulation.de/USA-Cities.html 
- Clean the data (around 30K cities, 3K counties, and 52states) to make every record in `neo/public/data/*.json` has a matching population data. Side note: Use the average population from counties for 880 cities and the average population from states for 76 counties to deal with the special cases. 

Setup procedure: 
- Ingest data into AsterixDB; data is in `noah/src/main/resources/population/adm`; example ingestion queries are in `noah/src/main/resources/population/sqlpp`
- register `dsStatePopulation`/`dsCountyPopulation`/`dsCityPopulation` using scripts in `scripts/register*.sh`

![nmlz1 mov](https://cloud.githubusercontent.com/assets/12385178/25105567/090d75b6-237a-11e7-9223-c5a70365e032.gif)
![nmlz2 mov](https://cloud.githubusercontent.com/assets/12385178/25105576/14a4a890-237a-11e7-9ac4-0847a12a38a8.gif)
![nmlz3 mov](https://cloud.githubusercontent.com/assets/12385178/25105579/188535f6-237a-11e7-968e-fc8eea5b8d58.gif)
![screen shot 2017-04-17 at 13 40 39](https://cloud.githubusercontent.com/assets/12385178/25105581/1d438282-237a-11e7-9779-5b09668f5018.jpg)
![screen shot 2017-04-17 at 13 40 51](https://cloud.githubusercontent.com/assets/12385178/25105587/20938bb2-237a-11e7-9df1-b1a890b7d376.jpg)

HotLemonJuice commented on neo/app/views/main.scala.html:
```diff
@@ -17,6 +17,7 @@
   <link rel="stylesheet" media="screen" href='@routes.Assets.versioned("lib/font-awesome/css/font-awesome.min.css")'>
   <link rel="stylesheet" media="screen" href='@routes.Assets.versioned("lib/bootstrap-vertical-tabs/bootstrap.vertical-tabs.min.css")'>
   <link rel='stylesheet' href='@routes.Assets.versioned("stylesheets/main.css")'>
+  <link href="https://gitcdn.github.io/bootstrap-toggle/2.2.2/css/bootstrap-toggle.min.css" rel="stylesheet">
```
I think I may need to use the Scala way to import library...

JavierJia replied:
have you figured this out? @vidhya567 can help on this.

JavierJia commented:
Nice :-). Maybe @luochen01 can give a review first?

luochen01 commented on neo/public/javascripts/map/controllers.js:
```diff
@@ -1,6 +1,9 @@
     $scope.result = {};
+    $scope.doNormalization = false;
+    $scope.upscaleFactor = 1000 * 1000;
+    $scope.upscaleFactorText = "M";
```
For scaleFactor and scaleFactorText, I think it's better to have a global config file or class for us to manage the demo more easily?

HotLemonJuice replied:
Good point. I wonder if we are going to use Scala or javascript

luochen01 replied:
I think it should be in js, since it's front-end related. Do we have some global config file or object for the TwitterMap right now? If the case, we can add these two options to it. If not, maybe we can have one...

luochen01 commented on neo/public/javascripts/map/controllers.js:
```diff
@@ -394,56 +399,94 @@ angular.module('cloudberry.map', ['leaflet-directive', 'cloudberry.common'])
+
       //FIXME: the code in county and city (and probably the state) levels are quite similar. Find a way to combine them.
-      if ($scope.status.logicLevel == "state" && $scope.geojsonData.state) {
+      if ($scope.status.logicLevel === "state" && $scope.geojsonData.state) {
```
I'm not familiar with JS... but why there is  ===?

HotLemonJuice replied:
=== means strict comparison. It's type-safe. In javascript, it's a preferred way to compare two variables when you don't want the type of the variables gives you unexpected results.

HotLemonJuice replied:
A good example taken from https://stackoverflow.com/questions/523643/difference-between-and-in-javascript:
```
0 == false   // true
0 === false  // false, because they are of a different type
1 == "1"     // true, automatic type conversion for value only
1 === "1"    // false, because they are of a different type
null == undefined // true
null === undefined // false
'0' == false // true
'0' === false // false```

luochen01 commented on neo/public/javascripts/common/services.js:
```diff
@@ -74,6 +74,18 @@ angular.module('cloudberry.common', [])
+      var targetUpperCase = targetLowerCase.charAt(0).toUpperCase() + targetLowerCase.substr(1);
+      return {
+        joinKey: [targetLowerCase],
+        dataset: "twitter.ds" + targetUpperCase + "Population",
```
It's better to have some meta object to map state/county/city to its dataset. Try to avoid string concatenation...

HotLemonJuice replied:
Do you mean to explicitly declare three separate query objects for state/county/city and use some conditions to map to them? And I wonder the reason to avoid using string concatenation. Is it because it's more error prone?

luochen01 replied:
Not that complex. You may define a map or a meta object to have mappings like state->statePopulation, county->countyPopulation etc. This mapping should also be resided in the global config object (if we have one).

luochen01 commented:
Sure, I made a few comments. However, the problem is I'm not familiar with javascript...

JavierJia commented:
Before jump into the code, in the screen shot, what does `Count:6.6M` mean in the legend?

HotLemonJuice commented:
I sent the dataset registering requests manually using terminal. I wrote `registerStatePopulation.sh` and two other similar shell scripts for that. Do I need to send the register in the front end?

HotLemonJuice commented:
Oops. I meaned to say 6.6/M. It means 6.6 tweets per million people. I will change it.

luochen01 commented:
I think maybe we need to come up with automatic way to register those datasets. Otherwise, other devs have to do it over and over again manually.

HotLemonJuice commented:
I have addressed all the issues @luochen01 mentioned. After discussing with @JavierJia , there is no need for the automatic data model registration for now. However, I did combine all the data model registration (ds_twitter and population) into one file: `./script/registerTwitterMapDataModel.sh`. It should make registration a bit easier.

luochen01 commented:
Saw it. And make sure to have clear instructions for those who want to try normalization on their local environment...

JavierJia commented on neo/public/javascripts/common/services.js:
```diff
@@ -1,5 +1,39 @@
+      }
+    };
+  })
+  .service('Asterix', function($http, $timeout, $location, cloudberryConfig) {
```
could you take this chance to rename this `Asterix`? it's been a while. It should be `cloudberry`.

JavierJia commented on neo/public/javascripts/map/controllers.js:
```diff
@@ -394,56 +397,94 @@ angular.module('cloudberry.map', ['leaflet-directive', 'cloudberry.common'])
-              d.properties.count = result[k].count;
+            if (result[k].city === d.properties.cityID) {
+              if($scope.doNormalization){
+                d.properties.count = result[k]['count'] / result[k]['population'] * cloudberryConfig.normalizationUpscaleFactor;
```
could you define a util function like `normalizeCount` to make the three places of the same code look better?

JavierJia commented on neo/public/javascripts/map/controllers.js:
```diff
@@ -464,19 +505,33 @@ angular.module('cloudberry.map', ['leaflet-directive', 'cloudberry.common'])
           }
-          if (d < 1000 * 1000) {
-            return (d / 1000).toString() + "K";
+          else if (d < 1000 * 1000) {
```
I think our convention for `if else` block is 
```
if {
} else  if {
} else {
}
```

JavierJia commented on noah/src/main/resources/population/crawler/population/population/items.py:
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
```
I can't comment on the empty file, is that `__init__.py` required? 
I'm sure we can remove the `*.pyc`.

HotLemonJuice replied:
I have removed `*.pyc`. `__init__.py` is required to make Python treat the directories as containing packages. Find more in this link: https://stackoverflow.com/questions/448271/what-is-init-py-for

JavierJia commented on noah/src/main/resources/population/crawler/population/population/spiders/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# This package will contain the spiders of your Scrapy project
```
remove those `*.pyc`

HotLemonJuice commented:
@JavierJia I have addressed all the issues you mentioned. And I have updated the documentation for instructions of setting up the environment for normalization.

chenlica commented:
Super!  Can we push it to the production system?

JavierJia commented:
Yes, we can. I will push it soon.
